[DDP] Allow await of custom buffer reduction in backward

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64515

For performance reasons, we would like to ensure that we can await
user collectives as part of custom buffer reduction in parallel to other work.
As a result, add support to return futures from custom buffer hooks and await
those futures at end of backwards pass.

Also added some docs to clarify how to use these APIs.

Differential Revision: [D30757761](https://our.internmc.facebook.com/intern/diff/D30757761/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

rohan-varma committed:
[DDP] Allow await of custom buffer reduction in backward

For performance reasons, we would like to ensure that we can await
user collectives as part of custom buffer reduction in parallel to other work.
As a result, add support to return futures from custom buffer hooks and await
those futures at end of backwards pass.

Also added some docs to clarify how to use these APIs.

Differential Revision: [D30757761](https://our.internmc.facebook.com/intern/diff/D30757761/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Allow await of custom buffer reduction in backward"


For performance reasons, we would like to ensure that we can await
user collectives as part of custom buffer reduction in parallel to other work.
As a result, add support to return futures from custom buffer hooks and await
those futures at end of backwards pass.

Also added some docs to clarify how to use these APIs.

Differential Revision: [D30757761](https://our.internmc.facebook.com/intern/diff/D30757761/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

zhaojuanmao commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -378,6 +378,14 @@ An enum-like class for built-in communication hooks: ``ALLREDUCE`` and ``FP16_CO
           },
           py::call_guard<py::gil_scoped_release>())
       .def("get_backward_stats", &::c10d::Reducer::get_backward_stats)
+      .def("_install_futures", [](::c10d::Reducer& reducer, const std::vector<std::shared_ptr<jit::PythonFutureWrapper>>& futs) {
```
nit: _install_post_backward_futures

zhaojuanmao commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -957,6 +957,10 @@ void Reducer::mark_bucket_ready(size_t bucket_index) {
 }
 
+void Reducer::install_futures(c10::List<c10::intrusive_ptr<c10::ivalue::Future>> futs) {
+    installed_futures_ = std::move(futs);
```
could we append futs here instead of moving?  it is possible that install_futures() was called multiple times and we need to wait all these installed futures?

rohan-varma replied:
Great idea!

zhaojuanmao commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -8010,10 +8008,26 @@ def buffer_comm_hook(ddp, named_buffers):
+                    # Note that when custom hooks return futures, this comparison is not expected to work when hook run location is pre-forward pass. This is because the hook does async communication
+                    # and forward pass modifies the buffer without appropriate synchronization. Therefore, if returning futures from custom buffer hooks, it is advised to set hook run location to post
+                    # forward.
+                    if return_futures and hook_run_location == hook_post_fwd:
```
hook_run_location does not have to be 'hook_post_fwd'?

rohan-varma replied:
It actually does, if you see the comments on L8017. Basically if we run the hook _before_ forward and return futures, then the user must perform appropriate synchronization in their forward pass for buffers to be equal. This is because the forward pass would use buffers without appropriate synchronization with the NCCL stream if we are returning futures. 

For the test we can manually change it to synchronize, but I think the best bet here is to document this and let users know they must explicitly synchronize if they want to use `return_futures` and `hook_run_location == hook_pre_fwd` (this is already documented on 1289).

rohan-varma committed:
Update on "[DDP] Allow await of custom buffer reduction in backward"


For performance reasons, we would like to ensure that we can await
user collectives as part of custom buffer reduction in parallel to other work.
As a result, add support to return futures from custom buffer hooks and await
those futures at end of backwards pass.

Also added some docs to clarify how to use these APIs.

Differential Revision: [D30757761](https://our.internmc.facebook.com/intern/diff/D30757761/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Allow await of custom buffer reduction in backward"


For performance reasons, we would like to ensure that we can await
user collectives as part of custom buffer reduction in parallel to other work.
As a result, add support to return futures from custom buffer hooks and await
those futures at end of backwards pass.

Also added some docs to clarify how to use these APIs.

Differential Revision: [D30757761](https://our.internmc.facebook.com/intern/diff/D30757761/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Allow await of custom buffer reduction in backward"


For performance reasons, we would like to ensure that we can await
user collectives as part of custom buffer reduction in parallel to other work.
As a result, add support to return futures from custom buffer hooks and await
those futures at end of backwards pass.

Also added some docs to clarify how to use these APIs.

Differential Revision: [D30757761](https://our.internmc.facebook.com/intern/diff/D30757761/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Allow await of custom buffer reduction in backward"


For performance reasons, we would like to ensure that we can await
user collectives as part of custom buffer reduction in parallel to other work.
As a result, add support to return futures from custom buffer hooks and await
those futures at end of backwards pass.

Also added some docs to clarify how to use these APIs.

Differential Revision: [D30757761](https://our.internmc.facebook.com/intern/diff/D30757761/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Allow await of custom buffer reduction in backward"


For performance reasons, we would like to ensure that we can await
user collectives as part of custom buffer reduction in parallel to other work.
As a result, add support to return futures from custom buffer hooks and await
those futures at end of backwards pass.

Also added some docs to clarify how to use these APIs.

Differential Revision: [D30757761](https://our.internmc.facebook.com/intern/diff/D30757761/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Allow await of custom buffer reduction in backward"


For performance reasons, we would like to ensure that we can await
user collectives as part of custom buffer reduction in parallel to other work.
As a result, add support to return futures from custom buffer hooks and await
those futures at end of backwards pass.

Also added some docs to clarify how to use these APIs.

Differential Revision: [D30757761](https://our.internmc.facebook.com/intern/diff/D30757761/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Allow await of custom buffer reduction in backward"


For performance reasons, we would like to ensure that we can await
user collectives as part of custom buffer reduction in parallel to other work.
As a result, add support to return futures from custom buffer hooks and await
those futures at end of backwards pass.

Also added some docs to clarify how to use these APIs.

Differential Revision: [D30757761](https://our.internmc.facebook.com/intern/diff/D30757761/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Allow await of custom buffer reduction in backward"


For performance reasons, we would like to ensure that we can await
user collectives as part of custom buffer reduction in parallel to other work.
As a result, add support to return futures from custom buffer hooks and await
those futures at end of backwards pass.

Also added some docs to clarify how to use these APIs.

Differential Revision: [D30757761](https://our.internmc.facebook.com/intern/diff/D30757761/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Allow await of custom buffer reduction in backward"


For performance reasons, we would like to ensure that we can await
user collectives as part of custom buffer reduction in parallel to other work.
As a result, add support to return futures from custom buffer hooks and await
those futures at end of backwards pass.

Also added some docs to clarify how to use these APIs.

Differential Revision: [D30757761](https://our.internmc.facebook.com/intern/diff/D30757761/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Allow await of custom buffer reduction in backward"


For performance reasons, we would like to ensure that we can await
user collectives as part of custom buffer reduction in parallel to other work.
As a result, add support to return futures from custom buffer hooks and await
those futures at end of backwards pass.

Also added some docs to clarify how to use these APIs.

Differential Revision: [D30757761](https://our.internmc.facebook.com/intern/diff/D30757761/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]
